web3/_utils/abi.py

```diff
@@ -342,6 +347,66 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
         return tuple()
 
 
+TUPLE_PREFIX = 'tuple'
+TUPLE_PREFIX_LEN = len(TUPLE_PREFIX)
+
+
+def _convert_abi_input(comp, arg):
+    """
+    Converts an argument ``arg`` corresponding to an ABI component ``comp``
+    into a plain value (for non-tuple components) or a properly converted and
+    ordered sequence (for tuple list components or tuple components).
+    """
+    abi_type_str = comp['type']
+
+    if not abi_type_str.startswith(TUPLE_PREFIX):
+        # Component is non-tuple.  Just return value.
+        return arg
+
+    # Component is tuple (possibly tuple list)...
+
+    array_dims = abi_type_str[TUPLE_PREFIX_LEN:]
+    if array_dims:
```

Comment:
Can we tighten this up at all?  Match against a regex or something?  This jumps out as a good way to end up with weird errors jumping out in the case that the ABI type string is malformed.

Reply:
Good idea.  I'll add that.